### PR TITLE
planner: refactor the IsolationReadEngines variable for plan cache

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/36598

Problem Summary: planner: refactor the IsolationReadEngines variable for plan cache

### What is changed and how it works?
1. Unimport the `IsolationReadEngines` in variables.
2. Use `GetIsolationReadEngines` to access the `IsolationReadEngines` values.
3. Use the `GetAvailableIsolationReadEngines4Plan` function to get the available isolation read engines. Not delete the tif lash engine first and add it later.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -1256,7 +1256,7 @@ func filterPathByIsolationRead(ctx sessionctx.Context, paths []*util.AccessPath,
 	if dbName.L == mysql.SystemDB {
 		return paths, nil
 	}
-	isolationReadEngines := ctx.GetSessionVars().GetIsolationReadEngines()
+	isolationReadEngines := GetAvailableIsolationReadEngines4Plan(ctx.GetSessionVars())
```
Comment (Reminiscent) [inline planner/core/planbuilder.go:1259]: Only here in the optimize phase need the `isolationReadEngines`. But it's not easy to get the `ast`. So we can not directly use `IsReadOnly` function to check. So we store the `IsReadyOnlyStmt` in advance.

File: planner/core/plan_cache.go [hunk 2]
```diff
@@ -220,15 +220,6 @@ func generateNewPlan(ctx context.Context, sctx sessionctx.Context, is infoschema
...
 	if prepared.UseCache && !stmtCtx.SkipPlanCache && !ignorePlanCache {
-		// rebuild key to exclude kv.TiFlash when stmt is not read only
-		if _, isolationReadContainTiFlash := sessVars.IsolationReadEngines[kv.TiFlash]; isolationReadContainTiFlash && !IsReadOnly(stmt, sessVars) {
-			delete(sessVars.IsolationReadEngines, kv.TiFlash)
-			if cacheKey, err = NewPlanCacheKey(sessVars, preparedStmt.StmtText, preparedStmt.StmtDB,
-				prepared.SchemaVersion, latestSchemaVersion); err != nil {
-				return nil, nil, err
-			}
-			sessVars.IsolationReadEngines[kv.TiFlash] = struct{}{}
-		}
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:231]: This part is useless here. We needn't handle the session variables in advance. Just use its original values.

File: server/conn_stmt.go [hunk 3]
```diff
@@ -216,9 +216,9 @@ func (cc *clientConn) handleStmtExecute(ctx context.Context, data []byte) (err e
 		// When the TiFlash server seems down, we append a warning to remind the user to check the status of the TiFlash
 		// server and fallback to TiKV.
 		prevErr := err
-		delete(cc.ctx.GetSessionVars().IsolationReadEngines, kv.TiFlash)
+		delete(cc.ctx.GetSessionVars().GetIsolationReadEngines(), kv.TiFlash)
```
Comment (qw4990) [inline server/conn_stmt.go:219]: It seems dangerous, `GetIsolationReadEngines()` should be read-only.

**Comment by hawkingrei** [comment]
because the CI upgrade go1.18.5. so I sync the code with master.

File: planner/core/cache.go [hunk 4]
```diff
@@ -149,13 +149,10 @@ func NewPlanCacheKey(sessionVars *variable.SessionVars, stmtText, stmtDB string,
 		lastUpdatedSchemaVersion: lastUpdatedSchemaVersion,
 		sqlMode:                  sessionVars.SQLMode,
 		timezoneOffset:           timezoneOffset,
-		isolationReadEngines:     make(map[kv.StoreType]struct{}),
+		isolationReadEngines:     sessionVars.GetIsolationReadEngines(),
```
Comment (qw4990) [inline planner/core/cache.go:152]: Should use `GetAvailableIsolationReadEngines4Plan()` here?

File: server/conn_stmt.go [hunk 5]
```diff
@@ -216,11 +216,9 @@ func (cc *clientConn) handleStmtExecute(ctx context.Context, data []byte) (err e
 		// When the TiFlash server seems down, we append a warning to remind the user to check the status of the TiFlash
 		// server and fallback to TiKV.
 		prevErr := err
-		delete(cc.ctx.GetSessionVars().IsolationReadEngines, kv.TiFlash)
-		defer func() {
-			cc.ctx.GetSessionVars().IsolationReadEngines[kv.TiFlash] = struct{}{}
-		}()
+		cc.ctx.GetSessionVars().StmtCtx.InTiFlashFallBack2TiKV = true
 		_, err = cc.executePreparedStmtAndWriteResult(ctx, stmt, args, useCursor)
+		cc.ctx.GetSessionVars().StmtCtx.InTiFlashFallBack2TiKV = false
```
Comment (qw4990) [inline server/conn_stmt.go:221]: Seem like we don't need to set it back.

File: sessionctx/variable/session.go [hunk 6]
```diff
@@ -1575,9 +1575,36 @@ func (s *SessionVars) GetSplitRegionTimeout() time.Duration {
...
+	return ok
+}
+
+// ContainTiFlashIsolationRead checks whether we can access the TiFlash read engines.
+func (s *SessionVars) ContainTiFlashIsolationRead() bool {
+	_, ok := s.isolationReadEngines[kv.TiFlash]
+	return ok
+}
+
 // GetIsolationReadEngines gets isolation read engines.
```
Comment (qw4990) [inline sessionctx/variable/session.go:1590]: Please add some comments to indicate it's read-only.

File: sessionctx/variable/session.go [hunk 7]
```diff
@@ -1575,9 +1575,36 @@ func (s *SessionVars) GetSplitRegionTimeout() time.Duration {
...
 }
 
+// ContainTiKVIsolationRead checks whether we can access the TiKV read engines.
+func (s *SessionVars) ContainTiKVIsolationRead() bool {
+	_, ok := s.isolationReadEngines[kv.TiKV]
+	return ok
+}
+
+// ContainTiFlashIsolationRead checks whether we can access the TiFlash read engines.
+func (s *SessionVars) ContainTiFlashIsolationRead() bool {
```
Comment (qw4990) [inline sessionctx/variable/session.go:1585]: How about merging them into one function like `ContainIsolationReadEngine(kv.TiKV/TiFlash)`?

File: session/session.go [hunk 8]
```diff
@@ -363,7 +363,7 @@ func (s *session) cleanRetryInfo() {
 	for i, stmtID := range retryInfo.DroppedPreparedStmtIDs {
 		if planCacheEnabled {
 			if i > 0 && preparedAst != nil {
-				plannercore.SetPstmtIDSchemaVersion(cacheKey, stmtText, preparedAst.SchemaVersion, s.sessionVars.IsolationReadEngines)
+				plannercore.SetPstmtIDSchemaVersion(cacheKey, stmtText, preparedAst.SchemaVersion, s.sessionVars.GetIsolationReadEngines())
```
Comment (qw4990) [inline session/session.go:366]: Should use `GetAvailableIsolationReadEngines()`?

**Review by qw4990 — APPROVED** [review]

File: sessionctx/variable/session.go [hunk 9]
```diff
@@ -1575,9 +1575,31 @@ func (s *SessionVars) GetSplitRegionTimeout() time.Duration {
...
 }
 
+// ContainSpecialIsolationRead checks whether we can access the special read engines.
+func (s *SessionVars) ContainSpecialIsolationRead(engineType kv.StoreType) bool {
+	_, ok := s.isolationReadEngines[engineType]
+	return ok
+}
+
 // GetIsolationReadEngines gets isolation read engines.
+// The return value is read-only.
```
Comment (xuyifangreeneyes) [inline sessionctx/variable/session.go:1585]: It seems that we create a copy of `s.isolationReadEngines` and modifying the copy doesn't affect `s.isolationReadEngines`. Why do we need to address that it is read-only?

Comment (xuyifangreeneyes) [reply sessionctx/variable/session.go:1585]: /cc @qw4990

Comment (Reminiscent) [reply sessionctx/variable/session.go:1585]: Maybe the comment is misleading. Just want to say the change in the outside will not affect the original value. I'll fix it.

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**